File: projects/06-PersistentSegmentTrees/implementations/PersistentSegmentTree.java
/* Persistent Segment Tree - Java implementation
 * Paul Lee kkz3uq
* Uses SUM function
 */

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Node {
    int value; // The SUM value of the segment
    Node left, right; // Left and right children

    public Node(int value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }

    public Node(int value, Node left, Node right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
}

public class PersistentSegmentTree {
    int size; // Size of the input array
    ArrayList<Node> versions; // Array to store root of each version of segment tree

    public PersistentSegmentTree(int[] arr) {
        if (arr.length == 0) {
            throw new IllegalArgumentException("Input array cannot be empty.");
        }
        this.size = arr.length; // Array length for boundary checks
        this.versions = new ArrayList<>();
        Node root = build(arr, 0, size - 1); // Build initial segment tree.
        versions.add(root); // Store initial version of the tree
    }

    // Build segment tree from the input array
    public Node build(int[] arr, int start, int end) {
        if (start == end) { // Leaf node
            return new Node(arr[start]);
        }
        int mid = (start + end) / 2; // Split array into two halves
        // Create left and right child and create new node with sum of the left and right child values.
        Node leftChild = build(arr, start, mid);
        Node rightChild = build(arr, mid + 1, end);
        return new Node(leftChild.value + rightChild.value, leftChild, rightChild);
    }

    // Create a new version of segment tree with updated value at specified index.
    private Node update(Node previousRoot, int start, int end, int index, int newValue) {
        if (start == end) { // Leaf node, create a new node with the updated value
            return new Node(newValue);
        }

        int mid = (start + end) / 2; // Split array into two halves
        Node leftChild, rightChild;
        // Recursively update left or right child based on index, while sharing unchanged nodes.
        if (index <= mid) {
            leftChild = update(previousRoot.left, start, mid, index, newValue);
            rightChild = previousRoot.right;
        } else {
            rightChild = update(previousRoot.right, mid + 1, end, index, newValue);
            leftChild = previousRoot.left;
        }
        // Create new node with updated value, which is sum of left and right child values.
        return new Node(leftChild.value + rightChild.value, leftChild, rightChild);
    }

    // Public method to update segment tree.
    public int update(int version, int index, int newValue) {
        if (version < 0 || version >= versions.size()) {
            throw new IllegalArgumentException("Version number out of bounds.");
        }
        if (index < 0 || index >= size) {
            throw new IllegalArgumentException("Index out of bounds.");
        }
        Node newRoot = update(versions.get(version), 0, size - 1, index, newValue);
        versions.add(newRoot); // Store new version of tree
        return versions.size() - 1; // Return index of new version
    }

    // Query sum of a range [ql, qr]
    private int query(Node root, int start, int end, int ql, int qr) {
        if (ql > end || qr < start) { // No overlap
            return 0;
        }
        if (ql <= start && end <= qr) { // Total overlap
            return root.value;
        }
        // Partial overlap, query both children and return sum of their results.
        int mid = (start + end) / 2;
        return query(root.left, start, mid, ql, qr) + query(root.right, mid + 1, end, ql, qr);
    }

    // Public method to query a specific version of segment tree.
    public int query(int version, int ql, int qr) {
        if (version < 0 || version >= versions.size()) {
            throw new IllegalArgumentException("Version number out of bounds.");
        }
        if (ql < 0 || qr >= size || ql > qr) {
            throw new IllegalArgumentException("Invalid Query Range.");
        }
        return query(versions.get(version), 0, size - 1, ql, qr);
    }

    // Helper method to get value at a specific index for a given version.
    public int getPoint(int version, int index) {
        return query(version, index, index); // Query the range [index, index] to get the point value
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        // Read N and Q (number of operations)
        String line = br.readLine();
        if (line == null) return;
        st = new StringTokenizer(line);
        int n = Integer.parseInt(st.nextToken());
        if (st.hasMoreTokens()) st.nextToken(); 

        // Read initial array values
        int[] arr = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        PersistentSegmentTree pst = new PersistentSegmentTree(arr);

        // Process operations
        String opLine;
        while ((opLine = br.readLine()) != null) {
            st = new StringTokenizer(opLine);
            if (!st.hasMoreTokens()) break;

            String op = st.nextToken();
            if (op.equals("Q")) {
                // Format: Q version left right
                int version = Integer.parseInt(st.nextToken());
                int left = Integer.parseInt(st.nextToken());
                int right = Integer.parseInt(st.nextToken());
                System.out.println(pst.query(version, left, right));
            } else if (op.equals("U")) {
                // Format: U version index newValue 
                int version = Integer.parseInt(st.nextToken());
                int index = Integer.parseInt(st.nextToken());
                int newValue = Integer.parseInt(st.nextToken());
                
                pst.update(version, index, newValue);
            }
        }
    }
}

File: projects/06-PersistentSegmentTrees/implementations/persistent_segment_tree.py
# Persistent Segment Tree
# Joshua Markle (cze9bj)
#
# This is really the same as a normal segment tree but stores multiple
# versions of the array as well. Each update creates a new version
#
# This segment tree uses the SUM function for simplicity
#
# The main optimization is structural sharing:
# when one index is updated, we only rebuild the nodes along the path
# from the root to the changed leaf, everything else is reused


# A node in the segment tree
class Node:
    # value: the aggregate value (SUM)
    # left:  pointer to the left child
    # right: pointer to the right child
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right


# The class implementation for the persistent segment tree
class PersistentSegmentTree:
    # n:        array length
    # versions: list of root nodes; each update appends to this list

    def __init__(self, arr):
        # No empty inputs
        if not arr:
            raise ValueError("Input array must not be empty!")

        # Store array length for bounds checking
        self.n = len(arr)

        # Build the first version of the tree from the initial array
        # versions[0] is the root of the initial segment tree
        self.versions = [self._build(arr, 0, self.n - 1)]

    # Build a normal segment tree
    def _build(self, arr, lo, hi):
        # lo and hi is the segment covered by the current node
        # Base case: if lo == hi, the range is singular and its a leaf node
        if lo == hi:
            return Node(arr[lo])

        # Split the segment into left and right halves
        mid = (lo + hi) // 2

        # Create child subtrees
        left = self._build(arr, lo, mid)
        right = self._build(arr, mid + 1, hi)

        # Merge: the current node stores the sum of the two children
        return Node(left.value + right.value, left, right)

    # Create a NEW version of the tree
    def _update(self, node, lo, hi, idx, new_value):
        # node:      root of the subtree from the old version
        # lo, hi:    segment represented by this subtree
        # idx:       array index to update
        # new_value: new value that should be written at arr[idx]

        # If at the leaf for idx, return a new leaf node
        if lo == hi:
            return Node(new_value)

        # Find midpoint to see if updated index is in the left or right segment
        mid = (lo + hi) // 2

        # Rebuild the child that contains idx; reuse other child
        if idx <= mid:
            new_left = self._update(node.left, lo, mid, idx, new_value)
            new_right = node.right
        else:
            new_left = node.left
            new_right = self._update(node.right, mid + 1, hi, idx, new_value)

        # Merge: create a new parent whose value is the sum of the new children
        return Node(new_left.value + new_right.value, new_left, new_right)

    # Public method to create a new version
    def update(self, version, idx, new_value):
        # version:   which existing version we want to update from
        # idx:       which array index to change
        # new_value: value to place at that index

        # Check that the version exists
        if version < 0 or version >= len(self.versions):
            raise IndexError("Bad version index!")

        # Check that the array index is valid
        if idx < 0 or idx >= self.n:
            raise IndexError("Index out of bounds!")

        # Make a new root by copying from the chosen version
        new_root = self._update(self.versions[version], 0, self.n - 1, idx, new_value)

        # Save as another version
        self.versions.append(new_root)

        # Return the index of the new version
        return len(self.versions) - 1

    # Get the range sum for [ql, qr]
    def _query(self, node, lo, hi, ql, qr):
        # Three cases:
        # 1. Total overlap: current segment is completely inside query range
        # 2. No overlap: current segment is completely outside query range
        # 3. Partial overlap: split and recurse

        # Total overlap
        if ql <= lo and hi <= qr:
            return node.value

        # No overlap
        if hi < ql or qr < lo:
            return 0

        # Partial overlap: query both children and add results
        mid = (lo + hi) // 2
        total = 0

        if node.left is not None:
            total += self._query(node.left, lo, mid, ql, qr)

        if node.right is not None:
            total += self._query(node.right, mid + 1, hi, ql, qr)

        return total

    # Public method for range sum queries
    def query(self, version, left, right):
        # version:     which historical version to query
        # left, right: inclusive query bounds

        # Validate version
        if version < 0 or version >= len(self.versions):
            raise IndexError("Bad version index!")

        # Validate range
        if left < 0 or right >= self.n or left > right:
            raise IndexError("Bad query range!")

        return self._query(self.versions[version], 0, self.n - 1, left, right)

    # Helper to get one value from one version
    def point_get(self, version, idx):
        return self.query(version, idx, idx)


# --- Take Input --- # 


# Input:
# n q
# array of n ints
# then q ops:
# U v i x (update -> new version)
# Q v l r (range sum)
# 0-based indexing, print result for each Q
def main():
    # Read n and q
    n, q = map(int, input().split())

    # Read initial array
    arr = list(map(int, input().split()))

    # Build tree
    pst = PersistentSegmentTree(arr)

    # Process commands
    outputs = []

    for _ in range(q):
        parts = input().split()
        cmd = parts[0]

        if cmd == "U":
            # U version index new_value
            version = int(parts[1])
            idx = int(parts[2])
            new_value = int(parts[3])
            pst.update(version, idx, new_value)

        elif cmd == "Q":
            # Q version left right
            version = int(parts[1])
            left = int(parts[2])
            right = int(parts[3])
            result = pst.query(version, left, right)
            outputs.append(str(result))

        else:
            raise ValueError("Invalid command")

    # Print results
    print("\n".join(outputs))


if __name__ == "__main__":
    main()
